pgmpy/models/MarkovChain.py

```diff
@@ -0,0 +1,299 @@
+        >>> from pgmpy.models import MarkovChain as MC
+        >>> model = MC()
+        >>> model.add_variable('grade', 3)
+        >>> grade_tm = {0: {0: 0.1, 1: 0.5, 2: 0.4}, 1: {0: 0.2, 1: 0.2, 2: 0.6 }, 2: {0: 0.7, 1: 0.15, 2: 0.15}}
+        >>> model.add_transition_model('grade', grade_tm)
+        """
+        # check if the transition model is valid
+        if not isinstance(transition_model, dict):
+            raise ValueError('Transition model must be a dict.')
+        if not set(transition_model.keys()) == set(range(self.cardinalities[variable])):
+            raise ValueError('Transitions must be defined for all states of variable {var}.'.format(var=variable))
+        for _, transition in transition_model.items():
+            if not isinstance(transition, dict):
+                raise ValueError('Each transition must be a dict.')
+            prob_sum = 0
+            for _, prob in transition.items():
+                if prob < 0 or prob > 1:
+                    raise ValueError('Transitions must represent valid probability weights.')
+                prob_sum += prob
+            if not np.allclose(prob_sum, 1):
+                raise ValueError('Transition probabilities must sum to 1.')
+
+        self.transition_models[variable] = transition_model
+
+    def sample(self, start_state=None, size=1):
+        """
+        Sample from the Markov Chain.
+
+        Parameters:
+        -----------
+        start_state: dict
+            representing the starting states of the variables.
+        size: int
+            number of samples to be generated.
+
+        Return Type:
+        ------------
+        pandas.DataFrame
+
+        Examples:
+        ---------
+        >>> from pgmpy.models import MarkovChain as MC
+        >>> model = MC(['intel', 'diff'], [2, 3])
+        >>> model.set_start_state({'intel': 0, 'diff': 2})
+        >>> intel_tm = {0: {0: 0.25, 1: 0.75}, 1: {0: 0.5, 1: 0.5}}
+        >>> model.add_transition_model('intel', intel_tm)
+        >>> diff_tm = {0: {0: 0.1, 1: 0.5, 2: 0.4}, 1: {0: 0.2, 1: 0.2, 2: 0.6 }, 2: {0: 0.7, 1: 0.15, 2: 0.15}}
+        >>> model.add_transition_model('diff', diff_tm)
+        >>> model.sample(size=5)
+           intel  diff
+        0      0     2
+        1      1     0
+        2      0     1
+        3      1     0
+        4      0     2
+        """
+        # check if the start state is valid
+        if start_state is not None and self._check_state(start_state):
+            self.state = start_state
+        elif start_state is None and self.state is None:
+            raise ValueError('Start state not set.')
+
+        sampled = DataFrame(index=range(size), columns=self.variables)
+        sampled.loc[0] = [self.state[var] for var in self.variables]
+        for i in range(size - 1):
+            for var in self.variables:
+                val = self.state[var]
+                next_val = sample_discrete(list(self.transition_models[var][val].keys()),
```

Comment:
You can put this list conversion out of the for loop to avoid computing it in each iteration.